src/lib/scheduler/node_queue_scheduler.cpp

```diff
@@ -118,4 +118,31 @@ void NodeQueueScheduler::schedule(std::shared_ptr<AbstractTask> task, NodeID pre
   auto queue = _queues[preferred_node_id];
   queue->push(task, static_cast<uint32_t>(priority));
 }
+
+void NodeQueueScheduler::_group_tasks(const std::vector<std::shared_ptr<AbstractTask>>& tasks) const {
+  // Adds predecessor/successor relationships between tasks so that only num_groups tasks can be executed in parallel.
+  // The optimal value of num_groups depends on the number of cores and the number of queries being executed
+  // concurrently. The current value has been found with a divining rod.
+  //
+  // Approach: Skip all tasks that already have predecessors or successors, as adding relationships to these could
+  // introduce cyclic dependencies. This ignores the preferred node of tasks. Again, this is far from perfect, but
```

Comment:
Reading this again makes me realize that the comment is slightly incorrect. It uses the *preferred node* information from the first task.

In all cases, this information will be the same across all tasks. Either it is unset or it uses the node that the current job operates on. There is no point in implementing anything here because it will not be used and cannot even be benchmarked.

What I could do is add an assertion that all preferred nodes are the same. Together with a comment that this is not really a requirement, but is supposed to draw the attention if we ever look deeper into NUMA things, this should cover all that we can do right now.

Reply:
Ok, the assertion sounds good.

Reply:
I should write those texts without using hypothetical grammar so that I can copy them right away :(